### PR TITLE
Institutional Metrics for Impact Script [ENG-1477]

### PR DESCRIPTION
**Pull request by UdayVarkhedkar**
## Purpose

Creates a management command to update institutional and institutional-user public and private project counts in Impact

## Changes

* Creation of `update_institution_project_counts` management command which records public and private project counts for institutions and institution-users.
* Creation of `UserInstitutionProjectCounts` and `InstitutionProjectCounts` metrics classes with functions to write and retrieve records from Impact. 

## QA Notes

These changes do not require data migrations and is an additive change. Testing for this ticket may have to be dev testing (outside of what I've already tested locally) by running the script and modifying the Impact indices in elasticsearch6. 

## Documentation

No external documentation needs to be updated.

## Side Effects

Possible side effects include increased load on servers at the time of day the management command is set to run.

## Ticket

[JIRA Ticket](https://openscience.atlassian.net/browse/ENG-1477)

### CODE REVIEW

**Comment by UdayVarkhedkar** [comment]
Example of user-institution records in Impact
```
{
	"took": 32,
	"timed_out": false,
	"_shards": {
		"total": 5,
		"successful": 5,
		"skipped": 0,
		"failed": 0
	},
	"hits": {
		"total": 6,
		"max_score": 1.0,
		"hits": [{
			"_index": "osf_userinstitutionprojectcounts_2020",
			"_type": "doc",
			"_id": "ur-vfXABLZtdJlCa5-bH",
			"_score": 1.0,
			"_source": {
				"timestamp": "2020-02-23T00:00:00",
				"user_id": "bd53u",
				"institution_id": "cos",
				"public_project_count": 8,
				"private_project_count": 6
			}
		}, {
			"_index": "osf_userinstitutionprojectcounts_2020",
			"_type": "doc",
			"_id": "u7-vfXABLZtdJlCa5-bj",
			"_score": 1.0,
			"_source": {
				"timestamp": "2020-02-21T00:00:00",
				"user_id": "bd53u",
				"institution_id": "cos",
				"public_project_count": 6,
				"private_project_count": 5
			}
		}, {
			"_index": "osf_userinstitutionprojectcounts_2020",
			"_type": "doc",
			"_id": "vL-vfXABLZtdJlCa6OYK",
			"_score": 1.0,
			"_source": {
				"timestamp": "2020-02-24T00:00:00",
				"user_id": "hy84n",
				"institution_id": "cos",
				"public_project_count": 19,
				"private_project_count": 14
			}
		}, {
			"_index": "osf_userinstitutionprojectcounts_2020",
			"_type": "doc",
			"_id": "vr-vfXABLZtdJlCa6OZM",
			"_score": 1.0,
			"_source": {
				"timestamp": "2020-02-21T00:00:00",
				"user_id": "hy84n",
				"institution_id": "cos",
				"public_project_count": 15,
				"private_project_count": 12
			}
		}, {
			"_index": "osf_userinstitutionprojectcounts_2020",
			"_type": "doc",
			"_id": "vb-vfXABLZtdJlCa6OYr",
			"_score": 1.0,
			"_source": {
				"timestamp": "2020-02-23T00:00:00",
				"user_id": "hy84n",
				"institution_id": "cos",
				"public_project_count": 17,
				"private_project_count": 13
			}
		}, {
			"_index": "osf_userinstitutionprojectcounts_2020",
			"_type": "doc",
			"_id": "ub-vfXABLZtdJlCa5-aW",
			"_score": 1.0,
			"_source": {
				"timestamp": "2020-02-24T00:00:00",
				"user_id": "bd53u",
				"institution_id": "cos",
				"public_project_count": 10,
				"private_project_count": 7
			}
		}]
	}
}
```

**Comment by UdayVarkhedkar** [comment]
Example of institution records in Impact
```
{
	"took": 21,
	"timed_out": false,
	"_shards": {
		"total": 5,
		"successful": 5,
		"skipped": 0,
		"failed": 0
	},
	"hits": {
		"total": 4,
		"max_score": 1.0,
		"hits": [{
			"_index": "osf_institutionprojectcounts_2020",
			"_type": "doc",
			"_id": "wr-zfXABLZtdJlCaTeYi",
			"_score": 1.0,
			"_source": {
				"timestamp": "2020-02-25T00:00:00",
				"institution_id": "cos",
				"public_project_count": 50,
				"private_project_count": 34
			}
		}, {
			"_index": "osf_institutionprojectcounts_2020",
			"_type": "doc",
			"_id": "wb-vfXABLZtdJlCa6eaP",
			"_score": 1.0,
			"_source": {
				"timestamp": "2020-02-21T00:00:00",
				"institution_id": "cos",
				"public_project_count": 38,
				"private_project_count": 26
			}
		}, {
			"_index": "osf_institutionprojectcounts_2020",
			"_type": "doc",
			"_id": "v7-vfXABLZtdJlCa6eY1",
			"_score": 1.0,
			"_source": {
				"timestamp": "2020-02-24T00:00:00",
				"institution_id": "cos",
				"public_project_count": 38,
				"private_project_count": 26
			}
		}, {
			"_index": "osf_institutionprojectcounts_2020",
			"_type": "doc",
			"_id": "wL-vfXABLZtdJlCa6eZv",
			"_score": 1.0,
			"_source": {
				"timestamp": "2020-02-23T00:00:00",
				"institution_id": "cos",
				"public_project_count": 38,
				"private_project_count": 26
			}
		}]
	}
}
```

File: osf/metrics.py [hunk 1]
```diff
@@ -175,3 +175,99 @@ class PreprintView(BasePreprintMetric):
...
+        )
+
+    @classmethod
+    def get_latest_user_institution_project_counts(cls, user, institution):
+        search = cls.search().filter('match', user_id=user._id).filter('match', institution_id=institution._id).sort('-timestamp')[:1]
+
+        try:
+            response = search.execute()
+        except NotFoundError:
+            # _get_relevant_indices returned 1 or more indices
```
Comment (Johnetordoff) [inline osf/metrics.py]: We use this exact span of code with these exact comments 5 fives in the repo, maybe make this a util? or remove 213-117, because I'm not sure this error can actually happen in this context.

File: osf/metrics.py [hunk 2]
```diff
@@ -175,3 +175,99 @@ class PreprintView(BasePreprintMetric):
 
 class PreprintDownload(BasePreprintMetric):
     pass
+
+
+class UserInstitutionProjectCounts(MetricMixin, metrics.Metric):
+    user_id = metrics.Keyword(index=True, doc_values=True, required=True)
+    institution_id = metrics.Keyword(index=True, doc_values=True, required=True)
+    # TODO: Uncomment following line and include department in metrics when available as a OSFUser attribute
+    # department = metrics.Keyword(index=True, doc_values=True, required=False)
```
[the remark this answers is not in the document]
Comment (UdayVarkhedkar) [reply osf/metrics.py]: That's a great point

File: osf/management/commands/update_institution_project_counts.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                institution=institution,
+                public_project_count=institution_public_projects_count,
+                private_project_count=institution_private_projects_count,
+                timestamp=now
+            )
+
+            for user in institution.osfuser_set.all():
+                user_public_project_count = Node.objects.get_nodes_for_user(
+                    user=user,
+                    permission=WRITE_NODE,
```
[the remark this answers is not in the document]
Comment (UdayVarkhedkar) [reply osf/management/commands/update_institution_project_counts.py]: Good call, you're correct that it is Read

File: osf/metrics.py [hunk 4]
```diff
@@ -175,3 +175,99 @@ class PreprintView(BasePreprintMetric):
...
+
+    class Index:
+        settings = {
+            'number_of_shards': 1,
+            'number_of_replicas': 1,
+            'refresh_interval': '1s',
+        }
+
+    @classmethod
+    def record_user_institution_project_counts(cls, user, institution, public_project_count, private_project_count, **kwargs):
```
[the remark this answers is not in the document]
Comment (UdayVarkhedkar) [reply osf/metrics.py]: You're right that it would work functionally but I think having an explicit record function for the class is still beneficial in making the class a bit more easy to understand and write code for.

File: osf/metrics.py [hunk 5]
```diff
@@ -175,3 +175,99 @@ class PreprintView(BasePreprintMetric):
...
+            institution_id=institution._id,
+            # TODO: Uncomment following line and include department in metrics when available as a OSFUser attribute
+            # department=user.department,
+            public_project_count=public_project_count,
+            private_project_count=private_project_count,
+            **kwargs
+        )
+
+    @classmethod
+    def get_latest_user_institution_project_counts(cls, user, institution):
```
[the remark this answers is not in the document]
Comment (UdayVarkhedkar) [reply osf/metrics.py]: It isn't an explicit requirement - I added them to complement the record methods and have more complete metrics classes. Good call with returning the document instead of the tuple - I'll change that.

File: osf/management/commands/update_institution_project_counts.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+from osf.metrics import InstitutionProjectCounts, UserInstitutionProjectCounts
+from osf.models import Institution, Node
+
+
+def update_institution_project_counts():
+    now = dt.datetime.now()
+
+    for institution in Institution.objects.all():
+
+        if institution.osfuser_set.exists() and institution.nodes.exists():
```
Comment (brianjgeiger) [inline osf/management/commands/update_institution_project_counts.py]: Drive-by: What if there are institutional users but no nodes? Probably could break this up into two `if` statements. Or is it even necessary? The query set counts would likely just be 0, and an empty base query set should run probably about as quickly as the `exists()` clauses, or close enough to.

Comment (UdayVarkhedkar) [reply osf/management/commands/update_institution_project_counts.py]: That's a good point - It is probably better if every possible institution-user possibility is retrievable, even if 0, from ES than requiring checks to be written in python when trying to retrieve the data, is that correct?

If so, I think removing that logic and allowing for 0 counts to be written is the best course of action.
